pkg/controller.v2/trainjob_controller.go

```diff
@@ -49,15 +54,70 @@ func (r *TrainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 	log := ctrl.LoggerFrom(ctx).WithValues("trainJob", klog.KObj(&trainJob))
-	ctrl.LoggerInto(ctx, log)
+	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling TrainJob")
+	if err := r.createOrUpdateObjs(ctx, &trainJob); err != nil {
+		return ctrl.Result{}, err
+	}
+	// TODO (tenzen-y): Do update the status.
 	return ctrl.Result{}, nil
 }
 
-func (r *TrainJobReconciler) SetupWithManager(mgr ctrl.Manager, runtimes map[string]runtime.Runtime) error {
+func (r *TrainJobReconciler) createOrUpdateObjs(ctx context.Context, trainJob *kubeflowv2.TrainJob) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Controller assumes the runtime existence has already verified in the webhook on TrainJob creation.
+	run := r.runtimes[runtimeRefToGroupKind(trainJob.Spec.RuntimeRef).String()]
+	objs, err := run.NewObjects(ctx, trainJob)
+	if err != nil {
+		return err
+	}
+	for _, obj := range objs {
+		var gvk schema.GroupVersionKind
+		if gvk, err = apiutil.GVKForObject(obj.DeepCopyObject(), r.client.Scheme()); err != nil {
+			return err
+		}
+		logKeysAndValues := []any{
+			"groupVersionKind", gvk.String(),
+			"namespace", obj.GetNamespace(),
+			"name", obj.GetName(),
```

Comment:
Do we set name and namespace for the build objects ?
E.g. here I can see that we only set the TypeMeta and Spec: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/runtime.v2/core/trainingruntime.go#L113-L119

Reply:
Yes, we do. Actual resource names are set in each plugins in the following:

JobSet: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/runtime.v2/framework/plugins/jobset/builder.go#L42-L43
PodGroup: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/runtime.v2/framework/plugins/coscheduling/coscheduling.go#L121-L122

Reply:
Why for JobSet plugin you want to have separate Builder, but for co-scheduling you create a PodSpec object directly under `Build()` API ?

Reply:
> Why for JobSet plugin you want to have separate Builder, but for co-scheduling you create a PodSpec object directly under Build() API ?

That is simply reason. There are so many fields and nested levels in the JobSet, but the PodGroup has a few field and a few nested level. For sure, we can provide the builder in the PodGroup as well. But, I guess that the PodGroup builder seems to be slightly overkill.

Reply:
Sounds good!